### PR TITLE
Change user for refusing leave

### PR DESCRIPTION
**Pull request by tschanzt**
@guewen as mentioned in the code for action_refuse: 
not sure about the copy but _remove_resource_leave searches through all leaves might be an issue with lots of data

also i added a test since i had trouble doing it manually.

### CODE REVIEW

File: connector_office_365_hr_leave/models/hr_leave.py [hunk 1]
```diff
@@ -50,3 +51,39 @@ def _validate_leave_request(self):
             if not leave.meeting_id and leave.meeting_id.user_id:
                 continue
             leave._office_365_push()
+
+    @api.multi
+    def action_refuse(self):
+        # not sure about the copy but _remove_resource_leave searches through
```
Comment (i-vyshnevska) [inline connector_office_365_hr_leave/models/hr_leave.py]: pls put a comment it's a copy of base method

File: connector_office_365_hr_leave/models/hr_leave.py [hunk 2]
```diff
@@ -50,3 +51,39 @@ def _validate_leave_request(self):
...
+                })
+            else:
+                holiday.write({
+                    'state': 'refuse',
+                    'second_approver_id': current_employee.id
+                })
+            # Delete the meeting
+            if holiday.meeting_id:
+                holiday.meeting_id.with_context(
+                    user=holiday.meeting_id.user_id
```
Comment (i-vyshnevska) [inline connector_office_365_hr_leave/models/hr_leave.py]: can we inject in context 'user' and call super().action_refuse()?

File: connector_office_365_hr_leave/tests/test_admin_can_delete_leaves.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+        self.user.office_365_persist_token(token)
+        self.leave = self.env['hr.leave'].create({
+            'employee_id': self.employee.id,
+            'date_from': datetime.now() + timedelta(days=1),
+            'date_to': datetime.now() + timedelta(days=2),
+            'number_of_days': 1,
+        })
+        self.leave.action_approve()
+        # we don't want to get a usererror here
+        self.leave.action_refuse()
```
Comment (i-vyshnevska) [inline connector_office_365_hr_leave/tests/test_admin_can_delete_leaves.py:83]: smth need to be checked here, you can check if meeting was deleted

File: connector_office_365_hr_leave/models/hr_leave.py (hunk 2 above)
Comment (tschanzt) [reply connector_office_365_hr_leave/models/hr_leave.py]: @i-vyshnevska no sadly not. if you refuse multiple leaves which  belong to different people the user must be the owner of the calendar entry. 
loop before calling super which lead to the comment above.

**Comment by tschanzt** [comment]
@i-vyshnevska @guewen i updated the pr.

File: connector_office_365_hr_leave/models/hr_leave.py [hunk 4]
```diff
@@ -50,3 +51,41 @@ def _validate_leave_request(self):
             if not leave.meeting_id and leave.meeting_id.user_id:
                 continue
             leave._office_365_push()
+
+    @api.multi
+    def action_refuse(self):
```
Comment (guewen) [inline connector_office_365_hr_leave/models/hr_leave.py:56]: We need to override a lot of things which is a pity...
Wouldn't this work as anyway the user we pass is always the user of the related meeting?

```
class HolidaysRequest(models.Model):

    @api.multi
    def action_refuse(self):
        return super(HolidaysRequest, self.with_context(o365_unlink_as_event_user=True).action_refuse()
    
class CalendarEvent(models.Model):

    @api.multi
    def unlink(self):
        if self.env.context.get('o365_unlink_as_event_user'):
            # loop on leaves, call super with event.user_id.id in 'user' context
```

Or even modify connector_office_365/models/calendar_event.py to check `o365_unlink_as_event_user` and set `user` as `event.user_id.id`

**Review by guewen — CHANGES_REQUESTED** [review]

File: connector_office_365_hr_leave/models/hr_leave.py (hunk 4 above)
Comment (tschanzt) [reply connector_office_365_hr_leave/models/hr_leave.py:56]: yes this way it would work. i'll change it.

**Comment by tschanzt** [comment]
@guewen updated.

File: connector_office_365_hr_leave/tests/test_admin_can_delete_leaves.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+        self.assertEqual(len(self.leave.meeting_id), 0)
```
Comment (i-vyshnevska) [inline connector_office_365_hr_leave/tests/test_admin_can_delete_leaves.py:84]: add check on amount of meetings before
add check that leave changed it's state

**Review by guewen — APPROVED** [review]

**Comment by guewen** [comment]
Thanks!

**Comment by tschanzt** [comment]
@i-vyshnevska updated.

**Review by i-vyshnevska — APPROVED** [review]
thanks, LG

**Review by i-vyshnevska — APPROVED** [review]

**Pull request merged**